### PR TITLE
Carry objects using the camera look direction

### PR DESCRIPTION
**Pull request by pcbeard**
Make Carriable subclassable:  add disable_collisions(), enable_collisions() overridable methods.

Add _integrate_forces(), _on_begin_dropping(), _on_end_dropping(). Reducing gravity and turning on contact monitoring and continuous collision detection to make dropping smoother.

Add Carriable.drop() method to drop objects intead of throwing them.

### CODE REVIEW

File: COGITO/Scripts/Carriable.gd [hunk 1]
```diff
@@ -19,12 +23,14 @@ func carry(player):
...
-		global_rotation = holder.carryable_position.global_rotation
+		# carry the object at the end of a stick where the user is looking.
+		global_position = camera.global_position + OFFSET - 2.5 * camera.get_global_transform().basis.z
+		global_rotation = holder.global_rotation
 #		set_global_transform(holder.carryable_position.get_global_transform())
 
 func hold():
-	$CollisionShape3D.set_disabled(true)
+	self.disable_collisions()
+	self.gravity_scale = 0.25
```
Comment (pcbeard) [inline COGITO/Scripts/Carriable.gd:33]: For some reason, I have to reduce gravity before dropping to ensure this takes effect. The value to adjust it by could certainly be an exported variable to tweak how this feels.

File: COGITO/Scripts/Carriable.gd [hunk 2]
```diff
@@ -19,12 +23,14 @@ func carry(player):
 
 func _process(_delta):
 	if is_being_carried:
-		global_position = holder.carryable_position.global_position
-		global_rotation = holder.carryable_position.global_rotation
+		# carry the object at the end of a stick where the user is looking.
+		global_position = camera.global_position + OFFSET - 2.5 * camera.get_global_transform().basis.z
+		global_rotation = holder.global_rotation
```
Comment (pcbeard) [inline COGITO/Scripts/Carriable.gd:28]: Using the camera's global position here makes carrying feel much smoother, like a steady-cam. This is of course a matter of taste and realism.

**Comment by pcbeard** [comment]
https://github.com/Phazorknight/Cogito/assets/793781/a14fc6dc-aec8-4a16-8ce2-8eeed3b3f7e5

**Comment by Phazorknight** [comment]
Looks pretty nice!

Took this as an opportunity to finally tackle reworking the carryable to actually keep collision active when carrying.
This is a bit of a drastic change and I know it's subjective, so I've kept it in the experimental branch for now. I've added some parameters for users to tweak to taste. Could probably easily implement a toggle to switch collisions off if needed.

Take a look here: **experimental / BETA 202401.7** 


https://github.com/Phazorknight/Cogito/assets/70389309/74208211-d9c8-4688-aa12-a2dded7c0386

**Comment by pcbeard** [comment]
Definitely an improvement. I think camera based carrying is awesome. Close mine if you're doing something better. But definitely let's keep this flexible so we can use with rigid body prefabs.

**Comment by pcbeard** [comment]
Maybe scroll wheel could control the the distance away from character (e.g. telescoping) and some kind of vertical trim vector (e.g. OFFSET) to change effective hold point.

**Comment by Phazorknight** [comment]
Closing this denying merging into main as this has been superceded by the new way carryables are handled.
I will turn the included carryables into classes soon so it's easier to implment custom behaviour if needed.

**Pull request closed without merging**